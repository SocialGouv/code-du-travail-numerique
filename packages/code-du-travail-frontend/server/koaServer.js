const Koa = require("koa");
const bodyParser = require("koa-bodyparser");
const helmet = require("koa-helmet");
const Router = require("koa-router");
const redirects = require("./redirects.json");

const IS_PRODUCTION_DEPLOYMENT =
  process.env.NEXT_PUBLIC_IS_PRODUCTION_DEPLOYMENT === "true";
const PROD_HOSTNAME = process.env.PROD_HOSTNAME || "code.travail.gouv.fr";
const AZURE_BASE_URL =
  process.env.AZURE_BASE_URL || "https://cdtnadmindev.blob.core.windows.net";

const dev = process.env.NODE_ENV !== "production";

const robotsDev = ["User-agent: *", "Disallow: /"].join("\n");
const robotsProd = [
  "User-agent: *",
  "Disallow: /assets/",
  "Disallow: /images/",
  "",
  `Sitemap: https://${PROD_HOSTNAME}/sitemap.xml`,
].join("\n");

async function getKoaServer({ nextApp }) {
  const nextHandler = nextApp.getRequestHandler();
  const server = new Koa();
  const router = new Router();

  const cspConfig = {
    directives: {
      defaultSrc: [
        "'self'",
        "*.travail.gouv.fr",
        "*.data.gouv.fr",
        "*.fabrique.social.gouv.fr",
      ],
      "font-src": ["'self'", "data:", "blob:"],
      frameSrc: [
        "'self'",
        "https://mon-entreprise.fr",
        "https://matomo.fabrique.social.gouv.fr",
        "*.dailymotion.com",
        "*.doubleclick.net",
      ],
      imgSrc: [
        "'self'",
        "data:",
        "*.fabrique.social.gouv.fr",
        "https://travail-emploi.gouv.fr",
        "https://mon-entreprise.fr",
        AZURE_BASE_URL,
        "*.doubleclick.net",
        "*.xiti.com",
      ],
      scriptSrc: [
        "'self'",
        "https://mon-entreprise.fr",
        "*.fabrique.social.gouv.fr",
        "https://cdnjs.cloudflare.com",
        "https://www.googletagmanager.com",
        "'unsafe-inline'",
        "*.doubleclick.net",
      ],
      styleSrc: ["'self'", "'unsafe-inline'"],
      ...(process.env.NEXT_PUBLIC_SENTRY_DSN && {
        reportUri: process.env.NEXT_PUBLIC_SENTRY_DSN,
      }),
    },
  };
  if (dev) {
    server.use(bodyParser());
    cspConfig.directives.scriptSrc.push("'unsafe-eval'");
  }
  server.use(helmet.contentSecurityPolicy(cspConfig));

  if (!IS_PRODUCTION_DEPLOYMENT) {
    server.use(async function (ctx, next) {
      ctx.set({ "X-Robots-Tag": "noindex, nofollow, nosnippet" });
      await next();
    });
  }

  router.get("/health", async (ctx) => {
    ctx.body = { status: "up and running" };
  });

  router.get("/robots.txt", async (ctx) => {
    ctx.type = "text/plain";
    ctx.body = IS_PRODUCTION_DEPLOYMENT ? robotsProd : robotsDev;
  });

  redirects.forEach(({ baseUrl, code, redirectUrl }) => {
    router.get(baseUrl, async (ctx) => {
      ctx.status = code;
      ctx.set("Location", redirectUrl);
      ctx.redirect(redirectUrl);
    });
  });

  router.all("(.*)", async (ctx) => {
    await nextHandler(ctx.req, ctx.res);
    ctx.respond = false;
  });

  server.use(async (ctx, next) => {
    ctx.status = 200;
    await next();
  });

  server.use(router.routes());
  return server;
}

module.exports = { getKoaServer };
