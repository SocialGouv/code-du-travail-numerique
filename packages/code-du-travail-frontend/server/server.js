// Import prerequisite packages
const next = require("next");
const { getKoaServer } = require("./koaServer");
/**
 * this env variable is use to target developpement / staging deployement
 * in order to block indexing bot using a x-robot-header and an appropriate robots.txt
 */
const IS_PRODUCTION_DEPLOYMENT =
  process.env.NEXT_PUBLIC_IS_PRODUCTION_DEPLOYMENT === "true";
const PORT = parseInt(process.env.FRONTEND_PORT, 10) || 3000;
const FRONTEND_HOST = process.env.FRONTEND_HOST || `http://localhost:${PORT}`;
const PROD_HOSTNAME = process.env.PROD_HOSTNAME || "code.travail.gouv.fr";
const NEXT_PUBLIC_SENTRY_DSN = process.env.NEXT_PUBLIC_SENTRY_DSN || "";
const PACKAGE_VERSION = process.env.VERSION || "";

const dev = process.env.NODE_ENV !== "production";

// Initialize NextJs instance and expose request handler
const nextApp = next({ dev });

nextApp.prepare().then(async () => {
  const server = await getKoaServer({
    nextApp,
  });

  server.listen(PORT, () => {
    //eslint-disable-next-line no-console
    console.log(`
  â€º Ready on ${FRONTEND_HOST}

  Environment:

  - process.env.NODE_ENV : ${process.env.NODE_ENV || "development"}
  - version: ${PACKAGE_VERSION}
  - is deployed on prod(${PROD_HOSTNAME}): ${IS_PRODUCTION_DEPLOYMENT}
  - sentry: ${NEXT_PUBLIC_SENTRY_DSN},

`);
  });

  this.server = server;
});

module.exports = nextApp;
