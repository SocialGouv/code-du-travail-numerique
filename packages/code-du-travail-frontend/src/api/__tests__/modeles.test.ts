/** @jest-environment node */

import request from "supertest";
import server from "nextjs-http-supertest";

describe("Modeles", () => {
  afterAll(() => {
    server.close();
  });

  it("should return all modeles", async () => {
    const res = await request(server).get("/api/modeles");
    expect(res.status).toBe(200);
    expect(res.body).toMatchSnapshot();
  });
});
