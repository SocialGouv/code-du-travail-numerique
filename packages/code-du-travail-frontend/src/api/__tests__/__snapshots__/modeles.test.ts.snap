// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modeles should return all modeles 1`] = `
[
  {
    "slug": "demande-de-rendez-vous-en-vue-dune-rupture-conventionnelle",
    "title": "Demande de rendez-vous en vue d’une rupture conventionnelle",
  },
  {
    "slug": "rupture-de-periode-dessai-a-linitiative-de-lemployeur",
    "title": "Rupture de période d’essai à l’initiative de l’employeur",
  },
  {
    "slug": "certificat-de-travail",
    "title": "Certificat de travail",
  },
]
`;
